Microsoft.Identity.Web/Client/TokenCacheProviders/Session/MSALPerUserSessionTokenCacheProvider.cs

```diff
@@ -104,23 +85,25 @@ private void UserTokenCacheBeforeWriteNotification(TokenCacheNotificationArgs ar
         /// <param name="args">Contains parameters used by the MSAL call accessing the cache.</param>
         private void UserTokenCacheAfterAccessNotification(TokenCacheNotificationArgs args)
         {
-            // if the access operation resulted in a cache update
+            // If the access operation resulted in a cache update
             if (args.HasStateChanged)
             {
-                string cacheKey = httpContextAccessor.HttpContext.User.GetMsalAccountId();
+                string cacheKey = _httpContextAccessor.HttpContext.Request.Headers["Authorization"];
```

Comment:
Again, this is a secret. Also did you test this? I suspect that the session cookie will be too large. We might want to keep only the signature of the access token?

Reply:
@jmprieur This is the http header of the incoming request, so not sure how it's related to the cookie. 
Yes, definitely tested. I wouldn't have done the PR w/out testing, but appreciate others testing as well. 

---
In reply to: [318914929](https://github.com/Azure-Samples/active-directory-dotnet-native-aspnetcore-v2/pull/93#discussion_r318914929) [](ancestors = 318914929)

Reply:
@jennyf19 : this is related to the cookie there are several possible implementatin of the cache serilization (See https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/Microsoft.Identity.Web#token-cache-serialization), and the session one uses session cookies.

Now it might be that in Web APIs, we only restrict the token cache implementation to InMemory? however:
- there is also the case of long running Web APIs (think of OneDrive which creates albums from your pictures long after you have stopped the OneDrive mobile app or left the Web site. I guess this one would still work with a different key?
- there is the scenario of hooks (for instance the Microsoft Graph hooks), where the identity of the user is passed as the iod and tid in a data structure.

 @yordan-msft, what are your recommendations:
- I understand that we can use the access token used to call the API as a key ? could/should we use something more coarse grain? (iod, tid, some particular claims?)
- Should Web APis use some serialiazers rather than others ? Is there such thing as a farm of Web APIs with a distributed token cache?

cc: @jennyf19 @henrik-me

Reply:
For webapi that does OBO, we should use access token for a key or key made up of tin/oid/device/mfa claims from the access token. 

Using only oil/tid is not sufficient.

Say if OBO access token (AT) was obtained using some webAPI access token, call it ATWeb, we should not give out the AT if webAPI is accessed with ac different access token - ATWeb2 - even if it is for same user, unless device, MFA etc. claims also match.
Easiest way to accomplish this is to use access token as cache key.

Regardless of what key is used, expiry of OBO access token is handled separately.

Reply:
Thanks Yordan for this update. The only issue I see is the size of the AT, but I guess it depends on the cache persistance later.

About the expiry I was probably not clear. The way we advise developers of confidential client apps to use MSAL.NET is to:
1) customize the token cache serialization
2) ensuring that the persistance mechanism persists one **token cache** per account (now per account/device/mfa=> AT used to call the Web API). This token cache is therefore only for one account, and contains several AT for downstream APIs. Now the persistance mechanism maintain a kind of dictionary of caches with keys was the account, and is now the AT. How do the app manage this dictionary? when do entries get removed? it seems to me that given the AT (used to call the Web API) has an expiry, this would be a good date at which to suppress the key (and therefore the cache) from the persistance layer?

More about the persistance of token caches in https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/master/Microsoft.Identity.Web/README.md#protected-web-apis-that-call-downstream-apis-on-behalf-of-a-user---startupcs

Reply:
if size is an issue, hash the AT and use the hash as key. This is a good practice anyway.

I see your point about cache. 
Let me replay - if we use an AT (or hash of it) as a cache key and the  AT expires at time X, then we should remove the cache entry after X, as the WebAPI will never get the same AT.
This makes sense to me. If we don't expire aggressively, the cache may fill up.